File: python/wildling/token.py
from __future__ import annotations

from typing import List, TypedDict


class TokenOptions(TypedDict, total=False):
    string: str
    startLength: int
    endLength: int
    variants: List[str]
    src: str


def _default_integer_option(option: object, fallback: int) -> int:
    return option if isinstance(option, int) and option >= 0 else fallback


class Token:
    def __init__(self, options: TokenOptions) -> None:
        self._src = options.get("src", "")
        self._start_length = _default_integer_option(options.get("startLength"), 1)
        self._end_length = _default_integer_option(options.get("endLength"), 1)
        self._variants = options.get("variants") or []
        self._count = 0
        for length in range(self._start_length, self._end_length + 1):
            self._count += len(self._variants) ** length

    def count(self) -> int:
        return self._count

    def src(self) -> str:
        return self._src

    def get(self, index: int) -> str:
        if index > self._count - 1 or index < 0:
            return ""

        if index == 0 and self._start_length == 0:
            return ""

        index_with_offset = index
        string_length = self._start_length
        for string_length in range(self._start_length, self._end_length + 1):
            offset_count = len(self._variants) ** string_length
            if index_with_offset < offset_count:
                break
            index_with_offset -= offset_count

        string_array: List[str] = []
        for _ in range(string_length):
            variant_index = index_with_offset % len(self._variants)
            index_with_offset //= len(self._variants)
            string_array.append(self._variants[variant_index])
        return "".join(string_array)


def create_token(options: TokenOptions) -> Token:
    return Token(options)

File: haskell/src/Wildling/Token.hs
module Wildling.Token
  ( Token(..)
  , TokenOptions(..)
  , createToken
  , tokenCount
  , tokenSrc
  , tokenGet
  ) where

data TokenOptions = TokenOptions
  { optString :: Maybe String
  , optStartLength :: Maybe Int
  , optEndLength :: Maybe Int
  , optVariants :: [String]
  , optSrc :: String
  }

data Token = Token
  { tokenSrc_ :: String
  , tokenStartLength :: Int
  , tokenEndLength :: Int
  , tokenVariants :: [String]
  , tokenCount_ :: Int
  }

defaultInteger :: Maybe Int -> Int -> Int
defaultInteger (Just n) _ | n >= 0 = n
defaultInteger _ fallback = fallback

powInt :: Int -> Int -> Int
powInt base expn
  | expn < 0 = 0
  | otherwise = go 1 expn
  where
    go acc 0 = acc
    go acc n = go (acc * base) (n - 1)

createToken :: TokenOptions -> Token
createToken options =
  let startLength = defaultInteger (optStartLength options) 1
      endLength = defaultInteger (optEndLength options) 1
      variants = optVariants options
      count =
        sum
          [ powInt (length variants) len
          | len <- [startLength .. endLength]
          ]
   in Token
        { tokenSrc_ = optSrc options
        , tokenStartLength = startLength
        , tokenEndLength = endLength
        , tokenVariants = variants
        , tokenCount_ = count
        }

tokenCount :: Token -> Int
tokenCount = tokenCount_

tokenSrc :: Token -> String
tokenSrc = tokenSrc_

tokenGet :: Token -> Int -> String
tokenGet token index
  | index > tokenCount_ token - 1 || index < 0 = ""
  | index == 0 && tokenStartLength token == 0 = ""
  | otherwise =
      let (stringLength, indexWithOffset) = findLength (tokenStartLength token) index
          variants = tokenVariants token
          n = length variants
       in buildString stringLength indexWithOffset n variants
  where
    findLength length_ idx
      | length_ > tokenEndLength token = (tokenEndLength token, idx)
      | otherwise =
          let offsetCount = powInt (length (tokenVariants token)) length_
           in if idx < offsetCount
                then (length_, idx)
                else findLength (length_ + 1) (idx - offsetCount)

    buildString 0 _ _ _ = ""
    buildString remaining idx n variants =
      let variantIndex = idx `mod` n
          nextIdx = idx `div` n
       in (variants !! variantIndex) ++ buildString (remaining - 1) nextIdx n variants
